keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1783 @@
+hardware and system configuration (device naming/numbering, IO scheduler
+configuration etc) which makes it very impractical to control from the Pod spec
+level. In order to hide this complexity the concept of blockio classes has been
+added to the container runtimes (CRI-O and containerd). A system administrator
+is able to configure blockio controller parameters on per-class basis and the
+classes are then made available for CRI clients. Following this model also
+provides a possible framework for the future improvements, for instance
+enabling class-based network or memory type prioritization of applications.
+
+Currently, there is no mechanism in Kubernetes to use these types of resources.
+CRI-O and containerd runtimes have support for RDT and blockio classes and they
+provide an bridge-gap user interface through special pod annotations. The goal
+is to get these types of resources first class citizens and properly supported
+in Kubernetes, providing visibility, a well-defined user interface, and
+permission controls.
+
+It seems necessary to support both container-level and pod-level QoS-class
+resources as independent concepts. Intel RDT (above) is per-container by
+design because of the hardware implementation (the control/class hiearchy is
+flat). Also, the current support for blockio is container-level only (it is not
+possible to configure pod sandbox-level cgroup parameters). However, having
+pod-level QoS-class resources makes it possible to implement support for
+sandbox-level blockio parameters. Other usage for pod sandbox-level QoS-class
+resources would be communicating the Kubernetes Pod QoS class from
+kubelet to the container runtime.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Make it possible to request QoS-class resources
+  - Support RDT class assignment of containers. This is already supported by
+    the containerd and CRI-O runtime and part of the OCI runtime-spec
+  - Support blockio class assignment of containers.
+  - Support Pod-level (sandbox-level) QoS-class resources
+- Make the API to support updating QoS-class resource assignment of running containers
+- Make the extensions flexible, enabling simple addition of other QoS-class
+  resource types in the future.
+- Make QoS-class resources opqaue (as possible) to the CRI client
+- Discovery of the available QoS-class resources
+- API changes to support updating Pod-level (sandbox-level) QoS-class resource
+  assignment of running pods
+- Resource status/capacity
+- Access control ([future work](#future-work))
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Interface or mechanism for configuring the QoS-class resources (responsibility of
+  the container runtime).
+- Enumerating possible (QoS-class) resource types or their detailed behavior
```

Comment:
My bad English skills. What I tried to say is not fixing in the API (or in the KEP) the resource types/names that are supported. I.e. when somebody comes up with a new type of (or vendor-specific) QoS resource it will be possible to implement it without needing to update the API.

Having said that, I think that we probably want to list "well-known" QoS resource types somewhere, e.g. in the CRI API.

I will re-phrase this

Reply:
Got it - enumerating != discovering. :)